### PR TITLE
Add TopNSigmaLogitsWarper and top_n_sigma generation config support

### LINKED ISSUE #39850: Support topNSigma sampling in `generate`
### Feature request

topNSigma is a sampling technique that guarantees that outputs are agnostic to temperature, while eliminating unrealistic tokens thereby reducing the chances of an illegitimate generation.

It's implementation centers around finding the std in the logits space and retaining the tokens whose logit scores are within the top-n standard deviations from the token corresponding to the max logit.

### Motivation

Their sampling technique shows improvements on a number of benchmarks when used with the Llama-3 family of models.

>  Extensive experiments across reasoning
and creative writing tasks demonstrate that our
method consistently outperforms existing approaches, with particularly significant improvements in high-temperature settings.

Link to the paper: https://aclanthology.org/2025.acl-long.528.pdf

### Your contribution

I'd love to create a PR for this!

### PR DESCRIPTION
**Pull request by ezylopx5**
## Problem

Transformers currently provides sampling filters such as top-k, top-p, min-p, and top-h, but does not include top-n-sigma sampling from "Top-nσ: Not All Logits Are You Need".

This makes it harder to use a temperature-invariant truncation rule based directly on pre-softmax logits.

Closes #39850.

## Root Cause

There is no logits warper implementing a dynamic threshold of:

`max(logits) - n_sigma * std(logits)`

and no corresponding `GenerationConfig` parameter routed through `GenerationMixin._get_logits_processor()`.

## Fix

This PR adds `TopNSigmaLogitsWarper` and integrates it into generation config and sampling flow.

### Code changes

- Add `TopNSigmaLogitsWarper` to generation logits processors.
- Add `top_n_sigma` field and validation in `GenerationConfig`.
- Add `top_n_sigma` sampling integration in `GenerationMixin`.
- Export `TopNSigmaLogitsWarper` from generation and top-level package init files.
- Add tests for behavior, validation, edge cases, and config serialization.

### Files changed

- `src/transformers/generation/logits_process.py`
- `src/transformers/generation/configuration_utils.py`
- `src/transformers/generation/utils.py`
- `src/transformers/generation/__init__.py`
- `src/transformers/__init__.py`
- `tests/generation/test_logits_process.py`
- `tests/generation/test_configuration_utils.py`

## Validation

Ran locally in editable install environment:

```bash
python -m pytest tests/generation/test_logits_process.py -k "top_n_sigma" -x -v
python -m pytest tests/generation/test_configuration_utils.py -k "top_n_sigma" -x -v
python -m pytest tests/generation/test_logits_process.py -x -v
python -m pytest tests/generation/test_configuration_utils.py -x -v
```

Results:

- `test_logits_process.py`: 50 passed
- `test_configuration_utils.py`: 34 passed, 5 skipped

Also ran style checks:

```bash
make style
```

## Risks

- Behavior is additive and opt-in (`top_n_sigma` defaults to `None`).
- Existing generation paths remain unchanged unless `top_n_sigma` is explicitly set.
- Includes validation for non-negative values to fail early on invalid input.

## Out of Scope

- Benchmarking quality/performance on large models.
- Interactions benchmarking vs top-p/min-p/top-h beyond functional correctness.

### CODE REVIEW

**Comment by ezylopx5** [comment]
Hi maintainers, CI is now green for this PR after adding the missing docs entry required by repository consistency.

This PR adds TopNSigmaLogitsWarper and top_n_sigma generation config support.

@gante @zucchini-nlp when you have a moment, could you please take a look? Thanks!

**Comment by Rocketknight1** [comment]
Please don't ask your code agent to find old issues and "fix" them!

**Comment by ezylopx5** [comment]
@Rocketknight1 thanks for the feedback, and sorry for the noise.You are right that I handled the process poorly. I moved too fast without prior discussion on the issue, and that made this look agent-first.
For transparency: I did use an agent for implementation support, but I personally reviewed the diff, ran the tests locally, fixed the failing repository-consistency check, and re-ran the relevant test files before opening the PR. So the technical validation was real on my side, but I understand that process and trust come first.
and i have closed the PR and deleted the branch. Going forward I will only propose changes after prior discussion on active issues and with smaller, clearly scoped contributions.
I appreciate the guidance and your time.
